File: packages/codegen/src/components/token-input.mjs
/**
 * SOLAR Token Input, beyond its IR: where MUI draws each layer and marks each state. Its shells are
 * files of their own, written by hand. One file per component, so adding one edits nothing shared;
 * `src/components/index.mjs` finds them.
 *
 * A field of entries it holds (owner decision 2026-09-24), each a SOLAR Tag, those past
 * \`maxVisible\` counted by a SOLAR Counter, and an input for the next: MUI's InputBase on the web,
 * an undecorated TextField in Flutter. Filled follows the entries and active the draft (the
 * overlay's `derive`).
 */

import { fieldResets } from './shared/field.mjs';

const P = 'SolarTokenInput';

export default {
  name: 'Token Input',
  mui: {
    // The shell draws every layer itself, each with a class of its own: the tokens are SOLAR Tags,
    // the input MUI's InputBase, in the field's row of tags.
    slots: 'drawn',
    resets: fieldResets('Token Input', {
      input: 'addItems',
      more: {
        // InputBase's own box around the input takes no part in the row's layout.
        [`& .${P}-words`]: { display: 'contents' },
        // The input takes the room the tokens leave, never less than a few characters.
        [`& .${P}--addItems.MuiInputBase-input`]: { minWidth: '4ch' },
        // The tokens keep their size; the row cuts off what does not fit (maxVisible counts it).
        [`& .${P}--tags`]: { minWidth: '0', overflow: 'hidden' },
        [`& .${P}--tags > *`]: { flexShrink: '0' },
      },
    }),
    // Hovered as the field is, and focused as the InputBase in it is; active (typing), filled,
    // read-only, in error and disabled by the shell's classes.
    states: {
      default: null,
      hover: `&:has(.${P}--field:hover)`,
      focus: `&:has(.${P}--field .Mui-focused)`,
      active: `&.${P}-active`,
      filled: `&.${P}-filled`,
      readonly: `&.${P}-readonly`,
      error: `&.${P}-error`,
      disabled: `&.${P}-disabled`,
    },
    overlaps: { focus: ['hover'] },
  },
  flutter: {},
  // How each platform reaches what the IR names, where not by its own name (src/shells/api.mjs).
  // A field's label is its `label`, as Text Input's is.
  // Flutter holds the draft in its controller.
  api: {
    react: { label: 'label' },
    // Flutter's word for a field that takes input (rule 5, owner decision 2026-09-25).
    flutter: { inputValue: 'controller', disabled: { not: 'enabled' } },
  },
};

File: packages/solar_flutter/variants/lib/src/token_input.dart
import 'package:flutter/material.dart';
import 'package:solar_flutter/solar_flutter.dart';

import 'probes.dart';

/// SolarTokenInput in one oracle variant: Figma's words, and as many entries as the variant
/// draws Tags, one more where it draws the Counter of the rest; typing Figma's words where it
/// draws it active; forced into a state through [states].
Widget buildTokenInput(
  Map<String, dynamic> v,
  WidgetStatesController states, [
  Map<String, dynamic>? _,
]) {
  final props = v['props'] as Map<String, dynamic>;
  final layers = v['layers'] as Map<String, dynamic>;
  final content = ((v['content'] as List?) ?? const []).cast<String>();
  bool drawn(String layer) =>
      layers[layer] != null &&
      (layers[layer] as Map<String, dynamic>)['hidden'] != true;
  final tags = ['tag', 'tag2'].where(drawn).length;
  return SolarTokenInput(
    size: enumNamed(SolarTokenInputSize.values, props['size'] as String),
    enabled: !(props['disabled'] as bool),
    error: props['error'] as bool,
    readonly: props['readonly'] as bool,
    label: 'Label',
    mandatory: true,
    helper: 'Helper text',
    value: [
      for (var i = 0; i < tags; i++) 'Label',
      if (drawn('counter')) 'More',
    ],
    maxVisible: tags,
    onChanged: (_) {},
    controller: TextEditingController(
      text: content.contains('inputValue') ? 'Add items…' : '',
    ),
    placeholder: 'Add items…',
    statesController: states,
  );
}
